### PR TITLE
add shape check for ARD hyperparameters to Kernel classes

### PR DESCRIPTION
**Pull request by st--**
Stationary kernels didn't check shape of lengthscales. If you have a non-ARD kernel, you should only be able to pass a single lengthscale. Shot myself in the foot because of this twice now. This PR adds checks (and tests for the checks).

### CODE REVIEW

**Comment by hughsalimbeni** [comment]
Could we also have a cast to float so that lengthscales=1 doesn't break it?

File: gpflow/kernels.py [hunk 1]
```diff
@@ -223,11 +223,16 @@ def __init__(self, input_dim, variance=1.0, lengthscales=None,
             else:
                 # accepts float or array:
                 lengthscales = lengthscales * np.ones(input_dim, dtype=settings.float_type)
+                if np.asarray(lengthscales).squeeze().shape != (input_dim,):
+                    raise ValueError("shape of lengthscales does not match input_dim")
```
Comment (awav) [inline gpflow/kernels.py]: Uppercase "Shape ..."

File: gpflow/kernels.py [hunk 2]
```diff
@@ -62,6 +62,37 @@ def __init__(self, input_dim, active_dims=None, name=None):
 
         self.num_gauss_hermite_points = 20
```
Comment (stefanosele) [inline gpflow/kernels.py]: I know that this is irrelevant to this PR but just saw that.
This is deprecated and should be removed. Kernels don't use `num_gauss_hermite_points` as far as I can tell.

File: gpflow/kernels.py [hunk 3]
```diff
@@ -60,7 +60,36 @@ def __init__(self, input_dim, active_dims=None, name=None):
...
+
+        Input:
+            value - scalar or array
+            ARD - None, False, or True. If None, infers ARD from shape of value
+        Output:
+            value - scalar if input_dim==1 or not ARD, array otherwise
+            ARD - False if input_dim==1 or not ARD, True otherwise
+        """
+        if ARD is None:
+            if np.asarray(value).squeeze().shape == ():
```
Comment (awav) [inline gpflow/kernels.py]: `ARD = np.asarray(value).squeeze().shape != ()`

File: gpflow/kernels.py [hunk 4]
```diff
@@ -60,7 +60,31 @@ def __init__(self, input_dim, active_dims=None, name=None):
             self.active_dims = np.array(active_dims, dtype=np.int32)
             assert len(active_dims) == input_dim
 
-        self.num_gauss_hermite_points = 20
+    def _validate_ard_shape(self, name, value, ARD=None):
```
Comment (awav) [inline gpflow/kernels.py]: What does `name` mean?

Comment (st--) [reply gpflow/kernels.py]: the name of the parameter, for sensible error messages

File: gpflow/kernels.py [hunk 5]
```diff
@@ -60,7 +60,31 @@ def __init__(self, input_dim, active_dims=None, name=None):
...
+        """
+        Validates the shape of a potentially ARD hyperparameter
+
+        :param value: A scalar or an array.
+        :param ARD: None, False, or True. If None, infers ARD from shape of value.
+        :return: Tuple (value, ARG), where _value_ is a scalar if input_dim==1 or not ARD, array otherwise.
```
Comment (st--) [inline gpflow/kernels.py]: ARG -> ARD

File: gpflow/kernels.py (hunk 4 above)
Comment (awav) [reply gpflow/kernels.py]: wouldn't it better to use kwargs then and restrict kwargs to one value? like:
```
def _validate_ard_shape(self, ARD: Optional[bool] = None, **kwargs: Dict[str, np.ndarray])
```

Comment (awav) [reply gpflow/kernels.py]: in any case, `name` doesn't have docstring

**Comment by st--** [comment]
ready to merge?

**Pull request merged**